### DISCUSSION TITLE
Entropy model in hific codebase

### DISCUSSION THREAD

**Discussion by mharshe** [tensorflow/compression#140] [General]
I'm looking at using the code from model.py to create a model that we can retrain. Our idea is to use tensorflow 2.0 keras layers where possible, but to port the code I see that the archs.py has this block of code:

class FactorizedPriorLayer(tf.keras.layers.Layer):
  """Factorized prior to code a discrete tensor."""

  def __init__(self):
    """Instantiate layer."""
    super(FactorizedPriorLayer, self).__init__(name="FactorizedPrior")
    self._entropy_model = tfc.EntropyBottleneck(
        name="entropy_model")

and it the class Hyperprior

self._side_entropy_model = FactorizedPriorLayer()

The EntropyBottleneck class was removed in commit https://github.com/tensorflow/compression/commit/89eefc85d8ececd149e7565da5ca9a4bed511e12 so I'm not sure how this code will run? Should I replace that side_entropy_model with NoisyDeepFactorized?

**Comment by jonaballe** [comment]
Hi, in general, the functionality of `EntropyBottleneck` is now found in `ContinuousBatchedEntropyModel` with an instance of `NoisyDeepFactorized` as the `prior` argument, and the equivalent of `GaussianConditional` is `ContinuousIndexedEntropyModel` with `NoisyGaussian` used in the `prior_fn` argument. But regarding the details in the HiFiC code base, @fab-jul is probably a better contact to ask.

**Reply by fab-jul** [reply]
The code should work with vanilla TFC classes, so using what you mentioned sounds good.

**Reply by mharshe** [reply]
I'm looking into changing the `GaussianConditional` to `ContinuousIndexedEntropyModel` but the input args are quite different.
The current code is : 
```
 conditional_entropy_model = tfc.GaussianConditional(
        latent_scales, scale_table, mean=latent_means,
        name="conditional_entropy_model")
```

whereas the `ContinuousIndexedEntropyModel` expects the arguments :  `prior_fn, index_ranges, parameter_fns, coding_rank`  I'm not sure what the arguments need to be changed to.

**Reply by jonaballe** [reply]
In this case, try [the location-scale entropy model](https://www.tensorflow.org/api_docs/python/tfc/entropy_models/LocationScaleIndexedEntropyModel).

**Reply by mharshe** [reply]
Ok, but still the question of the arguments remains -- GaussianConditional is given latent_scales & latent_means which are both the outputs of a Sequential block of 3 SignalConv2D layers. It isn't clear how these get translated to the arguments `num_scales`, `scale_fn` (which takes in indexes as a input parameter if I undertand right) and `coding_rank` of `LocationScaleIndexedEntropyModel`

**Reply by jonaballe** [reply]
Please take a look at this code example: https://github.com/tensorflow/compression/blob/master/models/bmshj2018.py#L157
`num_scales` and `scale_fn` represent the new way the `scale_table` is set up. `coding_rank` should be 3 since you presumably want to encode each HWC block of the NHWC tensor into its own bit sequence.
